datafusion/core/src/physical_plan/join_utils.rs

```diff
@@ -296,6 +299,154 @@ impl<T> Clone for OnceFut<T> {
+                JoinType::Full => {
+                    max(ij_cardinality, left_num_rows)
+                        + max(ij_cardinality, right_num_rows)
+                        - ij_cardinality
+                }
+                _ => unreachable!(),
+            };
+
+            Some(PartialJoinStatistics {
+                num_rows: cardinality,
+                // We don't do anything specific here, just combine the existing
+                // statistics which might yield subpar results (although it is
+                // true, esp regarding min/max). For a better estimation, we need
+                // filter selectivity analysis first.
+                column_statistics: all_left_col_stats
+                    .into_iter()
+                    .chain(all_right_col_stats.into_iter())
+                    .collect(),
+            })
+        }
+
+        JoinType::Semi => None,
+        JoinType::Anti => None,
+    }
+}
+
+/// Estimate the inner join cardinality by using the basic building blocks of
+/// column-level statistics and the total row count. This is a very naive and
+/// a very conservative implementation that can quickly give up if there is not
+/// enough input statistics.
+fn estimate_inner_join_cardinality(
+    left_num_rows: usize,
+    right_num_rows: usize,
+    left_col_stats: Vec<ColumnStatistics>,
+    right_col_stats: Vec<ColumnStatistics>,
+) -> Option<usize> {
+    // The algorithm here is partly based on the non-histogram selectivity estimation
+    // from Spark's Catalyst optimizer.
+
+    let mut join_selectivity = None;
+    for (left_stat, right_stat) in left_col_stats.iter().zip(right_col_stats.iter()) {
+        if (left_stat.min_value.clone()? > right_stat.max_value.clone()?)
+            || (left_stat.max_value.clone()? < right_stat.min_value.clone()?)
+        {
+            // If there is no overlap, then we can not accurately estimate
+            // the join cardinality. We could in theory use this information
+            // to point out the join will not produce any rows, but that would
+            // require some extra information (namely whether the statistics are
+            // exact). For now, we just give up.
+            return None;
+        }
+
+        let max_distinct = max(left_stat.distinct_count, right_stat.distinct_count);
+        if max_distinct > join_selectivity {
+            // Seems like there are a few implementations of this algorithm that implement
+            // exponential decay for the selectivity (like Hive's Optiq Optimizer). Needs
+            // further exploration.
+            join_selectivity = max_distinct;
+        }
+    }
+
+    // With the assumption that the smaller input's domain is generally represented in the bigger
+    // input's domain, we can estimate the inner join's cardinality by taking the cartesian product
+    // of the two inputs and normalizing it by the selectivity factor.
```

Comment:
One question: why use the selectivity factor(aka `max_distinct`) to normalize?

Reply:
Since this is an inner join, that means only the intersection on the join columns is going to be present in the result. In the conditions above, we have already verified that the ranges intersect so the number of distinct values (with a uniform distribution) is the most reliable way to actually find out the number of matching rows.

E.g. `a.col1` has `[1, 10000]` (all distinct) and `b.col1` has `[1, 1000]` (all distinct), for an inner join on `a.col1=b.col1`; we can estimate that the result cardinality as `10000 x 1000 / max(10000, 1000)` `=` `1000`. The same can be applied for distinct counts lower than the actual ranges (e.g. if `a.col1` had `1000` distinct values instead `10000` then there might be `10000` rows hence the normalization above ^)

Reply:
Got it, thanks a lot! @isidentical